
library(karyoploteR)
library(RColorBrewer)
custom.genome <- toGRanges(data.frame(chr=c('C1','C2','C3','C4','C5','C6','C7','C8','C9'), start=c(1,1,1,1,1,1,1,1,1), end=c(43764888,52886895,64984695,53719093,46902585,39822476,48366697,41758685,54679868)))

gffGenes <- import.gff('Brassica_oleracea.v2.1.34.chr_with_PAV.gff3')

#head(gffGenes)
# These files were generated by RGAugury
gffNBS <- import.gff('Brassica.CViT.NBS.blue.txt')
gffRLK <- import.gff('Brassica.CViT.RLK.green.txt')
gffRLP <- import.gff('Brassica.CViT.RLP.orange.txt')
gffTMCC <- import.gff('Brassica.CViT.TMCC.black.txt')

#cols <- brewer.pal(6, 'Dark2')
gffGenes <- gffGenes[gffGenes$type == 'gene']
png('All_densitiesGenes.png', width=480*3, height=480*3)
kp <- plotKaryotype(genome = custom.genome, plot.type=4, ideogram.plotter = NULL, cex=1.4)
pp <- getDefaultPlotParams(plot.type=4)
pp$ideogramlateralmargin <- 0.009

kp <- plotKaryotype(genome = custom.genome, plot.type=4, cex=1.4, plot.params=pp)
kpAddBaseNumbers(kp, cex=1.4)

kpDataBackground(kp, data.panel = 1, r0=0, r1=0.2)
kp <- kpPlotDensity(kp, gffGenes, r0=0, r1=0.2, col=cols[1])#, window.size=10e6)
kpAddLabels(kp, labels='All\nGenes', pos=1, r0=0, r1=0.2, cex=1.4, label.margin=0.02)

kpDataBackground(kp, data.panel = 1, r0=0.23, r1=0.4)
kp <- kpPlotDensity(kp, gffNBS, r0=0.23, r1=0.4, col=cols[2])
kpAddLabels(kp, labels='NBS', pos=1, r0=0.23, r1=0.4, cex=1.4, label.margin=0.02)

kpDataBackground(kp, data.panel = 1, r0=0.43, r1=0.6)
kp <- kpPlotDensity(kp, gffRLK, r0=0.43, r1=0.6, col=cols[3])
kpAddLabels(kp, labels='RLK', pos=1, r0=0.43, r1=0.6, cex=1.4, label.margin=0.02)

kpDataBackground(kp, data.panel = 1, r0=0.63, r1=0.8)
kp <- kpPlotDensity(kp, gffRLP, r0=0.63, r1=0.8, col=cols[4])
kpAddLabels(kp, labels='RLP', pos=1, r0=0.63, r1=0.8, cex=1.4, label.margin=0.02)

kpDataBackground(kp, data.panel = 1, r0=0.73, r1=0.9)
kp <- kpPlotDensity(kp, gffTMCC, r0=0.73, r1=0.9, col=cols[5])
kpAddLabels(kp, labels='TMCC', pos=1, r0=0.73, r1=0.9, cex=1.4, label.margin=0.02)

sign.genes <- gffGenes[as.numeric(gffGenes$PAV)>1]
kp <- kpPoints(kp, data=sign.genes, y=as.numeric(sign.genes$PAV), ymin=0.0, ymax=100.0, r0=0.93, r1=1.1, col=cols[6])
kpDataBackground(kp, data.panel = 1, r0=0.83, r1=1)
kpAddLabels(kp, labels='Variable\ngenes', pos=1, r0=0.83, r1=1, cex=1.4, label.margin=0.02)
kp <- kpPlotDensity(kp, sign.genes, r0=0.83, r1=1, col=cols[6])
dev.off()


library(regioneR)

print('NBS')
pt <- permTest(A=sign.genes, B=gffNBS, ntimes=500, genome=custom.genome, randomize.function=resampleRegions, evaluate.function=numOverlaps, force.parallel=F, universe=gffGenes)
pt
png('NBS_pt.png')
plot(pt)
dev.off()
lz <- localZScore(pt=pt, A=sign.genes, B=gffNBS)
png('NBS_lz.png')
plot(lz, pars=list(xlim=c(-1, 25)))
print(par('usr'))
dev.off()
print('RLK')
pt <- permTest(A=sign.genes, B=gffRLK, ntimes=500, genome=custom.genome, randomize.function=resampleRegions, evaluate.function=numOverlaps, force.parallel=F, universe=gffGenes)
pt

png('RLK_pt.png')
plot(pt)
dev.off()
lz <- localZScore(pt=pt, A=sign.genes, B=gffRLK)
png('RLK_lz.png')
plot(lz, pars=list(xlim=c(-1, 25)))
dev.off()

print('RLP')
pt <- permTest(A=sign.genes, B=gffRLP, ntimes=500, genome=custom.genome, randomize.function=resampleRegions, evaluate.function=numOverlaps, force.parallel=F, universe=gffGenes)
pt

png('RLP_pt.png')
plot(pt)
dev.off()
lz <- localZScore(pt=pt, A=sign.genes, B=gffRLP)
png('RLP_lz.png')
plot(lz, pars=list(ylim=c(-1, 25)))
dev.off()

print('TMCC')
pt <- permTest(A=sign.genes, B=gffTMCC, ntimes=500, genome=custom.genome, randomize.function=resampleRegions, evaluate.function=numOverlaps, force.parallel=F, universe=gffGenes)
pt

png('TMCC_pt.png')
plot(pt)
dev.off()
lz <- localZScore(pt=pt, A=sign.genes, B=gffTMCC)
png('TMCC_lz.png')
plot(lz, pars=list(ylim=c(-1, 25)))
dev.off()


sign.genes <- sign.genes[sign.genes$type == 'gene']

te.genes <- import.gff('Order1_TO1000_Mito_Chloro_all_200bp_filtered_contaminants_marked.fasta.out_only_C.gff')
print(te.genes)

print('checking TE with NBS')

pt <- permTest(B=te.genes, A=gffNBS, ntimes=500, genome=custom.genome, randomize.function=resampleRegions, evaluate.function=meanDistance, universe=gffGenes)
pt
png('TE_NBS_pt.png')
plot(pt)
dev.off()
lz <- localZScore(pt=pt, B=te.genes, A=gffNBS)
png('NBS_TE_lz.png')
plot(lz, pars=list(xlim=c(-1, 25)))
dev.off()

pt <- permTest(B=te.genes, A=gffRLK, genome=custom.genome, randomize.function=resampleRegions, evaluate.function=meanDistance, universe=gffGenes, ntimes=500)
print('TE vs RLK')
pt
png('TE_RLK_pt.png')
plot(pt)
dev.off()
lz <- localZScore(pt=pt, B=te.genes, A=gffRLK)
png('RLK_TE_lz.png')
plot(lz, pars=list(xlim=c(-1, 25)))
dev.off()

pt <- permTest(B=te.genes, A=gffRLP, genome=custom.genome, randomize.function=resampleRegions,evaluate.function=meanDistance, universe=gffGenes, ntimes=500)
print('TE vs RLP')
pt
png('TE_RLP_pt.png')
plot(pt)
dev.off()
lz <- localZScore(pt=pt, B=te.genes, A=gffRLP)
png('RLP_TE_lz.png')
plot(lz, pars=list(xlim=c(-1, 25)))
dev.off()

pt <- permTest(B=te.genes, A=gffTMCC, genome=custom.genome, randomize.function=resampleRegions,evaluate.function=meanDistance, universe=gffGenes, ntimes=500)
print('TE vs TMCC')
pt
png('TE_TMCC_pt.png')
plot(pt)
dev.off()
lz <- localZScore(pt=pt, B=te.genes, A=gffTMCC)
png('TMCC_TE_lz.png')
plot(lz, pars=list(xlim=c(-1, 25)))
dev.off()

print('checking TE with PAV')

pt <- permTest(B=te.genes, A=sign.genes, ntimes=500, genome=custom.genome, randomize.function=resampleRegions, evaluate.function=meanDistance, universe=gffGenes)
pt
png('TE_PAV_pt.png')
plot(pt)
dev.off()
lz <- localZScore(pt=pt, B=te.genes, A=gffNBS)
png('PAV_TE_lz.png')
plot(lz, pars=list(xlim=c(-1, 25)))
dev.off()

sessionInfo()
